File: stk_to_piet.py
import sys
import png

from generate_number import * # optimize number

def stk_to_piet(i_file, o_file, optim=True):
    # ⚪

    # ❤🧡💛💚💙💜
    # 🔴🟠🟡🟢🔵🟣
    # 🟥🟧🟨🟩🟦🟪

    # ⚫

    colors = [["❤","🧡","💛","💚","💙","💜"],
              ["🔴","🟠","🟡","🟢","🔵","🟣"],
              ["🟥","🟧","🟨","🟩","🟦","🟪"]]

    rev_map = {
        "❤": (0,0),
        "🔴": (0,1),
        "🟥": (0,2),
        "🧡": (1,0),
        "🟠": (1,1),
        "🟧": (1,2),
        "💛": (2,0),
        "🟡": (2,1),
        "🟨": (2,2),
        "💚": (3,0),
        "🟢": (3,1),
        "🟩": (3,2),
        "💙": (4,0),
        "🔵": (4,1),
        "🟦": (4,2),
        "💜": (5,0),
        "🟣": (5,1),
        "🟪": (5,2)}

    stack_optimizer = StackOptimizer()
    
    def make_block(block):
        output = ["🔴"]
        previous_c = 0
        previous_r = 1
        for inp in block:
            cmd = inp.split()
            match cmd[0]:
                case "push":
                    for _ in range(int(cmd[1])-1):
                        output.append(colors[previous_r][previous_c])
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "pop":
                    previous_r = (previous_r - 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "add":
                    previous_c = (previous_c + 1) % 6
                    output.append(colors[previous_r][previous_c])
                case "sub":
                    previous_c = (previous_c + 1) % 6
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "mul":
                    previous_c = (previous_c + 1) % 6
                    previous_r = (previous_r - 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "div":
                    previous_c = (previous_c + 2) % 6
                    output.append(colors[previous_r][previous_c])
                case "mod":
                    previous_c = (previous_c + 2) % 6
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "not":
                    previous_c = (previous_c + 2) % 6
                    previous_r = (previous_r - 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "greater":
                    previous_c = (previous_c + 3) % 6
                    output.append(colors[previous_r][previous_c])
                case "dup":
                    previous_c = (previous_c + 4) % 6
                    output.append(colors[previous_r][previous_c])
                case "roll":
                    previous_c = (previous_c + 4) % 6
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "inN":
                    previous_c = (previous_c + 4) % 6
                    previous_r = (previous_r - 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "inC":
                    previous_c = (previous_c + 5) % 6
                    output.append(colors[previous_r][previous_c])
                case "outN":
                    previous_c = (previous_c + 5) % 6
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "outC":
                    previous_c = (previous_c + 5) % 6
                    previous_r = (previous_r - 1) % 3
                    output.append(colors[previous_r][previous_c])
                case "branch":
                    # pointer
                    previous_c = (previous_c + 3) % 6
                    previous_r = (previous_r + 1) % 3
                    output.append(colors[previous_r][previous_c])

                    return output, [cmd[1], cmd[2]]
                case "goto":
                    return output, [cmd[1]]
                case "term":
                    return output, ["term"]
                case default:
                    print("Invalid cmd:", cmd)
        return output, []

    inp_lines = []
    with open(i_file, 'r') as f:
        inp_lines = f.readlines()
    lines = []
    for x in inp_lines:
        if "#" in x:
            lines.append(x[:x.find("#")])
        else:
            lines.append(x)

    blocks = dict()
    for l in [x.split("\n") for x in "".join(lines).split("\n\n")]:
        cmds = []
        for x in l[1:]:
            cmd = x.split()
            if len(cmd) == 2 and cmd[0] == "push":
                push_val = int(cmd[1])
                cmds = cmds + (stack_optimizer.optimize_number(int(cmd[1])) if optim else (["push "+str(push_val)] if push_val > 0 else ["push 1", "not", "push "+str(push_val), "sub"]))
            else:
                if len(x) > 0:
                    cmds.append(x)
        blocks[l[0].split()[1]] = make_block(cmds)
    blocks["term"] = ([], [])

    print ("Reading from:", i_file)
    print ("Saving to:", o_file)

    with open(o_file, 'w') as f:
        final_output = []

        push_max = len(blocks)*2
        pre = len(blocks)*2

        final_output.append("⚪"*(pre+push_max+1))
        for i in range(len(blocks)+1):
            final_output.append("⚪"*(pre+push_max+1))

        final_output[1] = final_output[1][:pre-1] + "⚫⚫" + final_output[i+1][pre-1+2:]

        for i in range(1,len(blocks)):
            j = i
            final_output[i+1] = final_output[i+1][:pre-1-(j*2)] + "⚫" + final_output[i+1][pre-1-(j*2)+1:]
            final_output[i+1] = final_output[i+1][:pre-1+(i*2)] + "⚫" + final_output[i+1][pre-1+(i*2)+1:]

        final_output[len(blocks)+1] = final_output[len(blocks)+1][:pre-1+(len(blocks)*2)] + "⚫" + final_output[i+1][pre-1+(len(blocks)*2)+1:]

        key_to_index = dict()
        prev_size = 0
        prev_clearance = 0
        for i, l in enumerate(sorted(blocks, key=lambda x: len(blocks[x][0]))):
            key_to_index[l] = i
            curr_size = len(blocks[l][0])

            clearance = 2
            if curr_size >= 2:
                ac, ar = rev_map[blocks[l][0][-1]]
                bc, br = rev_map[blocks[l][0][-2]]
                clearance = 3 # if ((ac - bc) % 6 == 3 and (ar - br) % 3 == 1) else 2
            else:
                clearance = 2

            if curr_size <= prev_size + clearance:
                blocks[l] = (["⚪"] * (prev_size + (0 if clearance == prev_clearance else -1) + clearance - curr_size) + blocks[l][0], blocks[l][1])

            prev_size = len(blocks[l][0])
            prev_clearance = clearance


        mid_section = []
        max_block_len = max([len(blocks[l][0]) for l in blocks])
        print (max_block_len)
        for xi in range(max_block_len + 2):
            xvals = ["⚪" for _ in range(pre + push_max + 1)]
            mid_section = ["".join(xvals)] + mid_section

        final_output_offset = len(final_output)
        final_output = final_output + mid_section

        for l in blocks:
            i = key_to_index[l]
            for xi, x in enumerate(blocks[l][0]):
                final_output[xi + final_output_offset] = final_output[xi+final_output_offset][:pre + (i*2)] + x + final_output[xi+final_output_offset][pre + (i*2) + 1:]
            if len(blocks[l][1]) == 1:
                j = key_to_index[blocks[l][1][0]]
                final_output[len(blocks[l][0])-1+final_output_offset] = final_output[len(blocks[l][0])-1+final_output_offset][:pre-(j * 2+2)] + "⚫" + final_output[len(blocks[l][0])-1+final_output_offset][pre-(j * 2+2)+1:]
                final_output[len(blocks[l][0])+final_output_offset] = final_output[len(blocks[l][0])+final_output_offset][:pre+(i*2)] + "⚫" + final_output[len(blocks[l][0])+final_output_offset][pre+i*2+1:]
            if len(blocks[l][1]) == 2:
                j = key_to_index[blocks[l][1][0]]
                final_output[len(blocks[l][0])-1+final_output_offset] = final_output[len(blocks[l][0])-1+final_output_offset][:pre-(j * 2+2)] + "⚫" + final_output[len(blocks[l][0])-1+final_output_offset][pre-(j * 2+2)+1:]

                j = key_to_index[blocks[l][1][1]]
                final_output[len(blocks[l][0])+final_output_offset] = final_output[len(blocks[l][0])+final_output_offset][:pre-(j * 2+2)] + "⚫" + final_output[len(blocks[l][0])+final_output_offset][pre-(j * 2+2)+1:]
                final_output[len(blocks[l][0])+1+final_output_offset] = final_output[len(blocks[l][0])+1+final_output_offset][:pre+i*2] + "⚫" + final_output[len(blocks[l][0])+1+final_output_offset][pre+i*2+1:]

        i = key_to_index["term"]
        final_output.append("⚪" * (pre+(i*2)-2) + "⚪⚫⚪⚫⚪" + "⚪" * (push_max-(i*2)-2))
        final_output.append("⚪" * (pre+(i*2)-2) + "⚫🔴🔴🔴⚫" + "⚪" * (push_max-(i*2)-2))
        final_output.append("⚪" * (pre+(i*2)-2) + "⚪⚫🔴⚫⚪" + "⚪" * (push_max-(i*2)-2))
        final_output.append("⚪" * (pre+(i*2)-2) + "⚪⚪⚫⚪⚪" + "⚪" * (push_max-(i*2)-2))

        final_output.append("⚪" * (pre  + push_max + 1))
        main_goto = key_to_index["main"]
        final_output[-1] = final_output[-1][:pre-(main_goto * 2+2)] + "⚫" + final_output[-1][pre-(main_goto * 2+2)+1:]

        while (final_output[len(blocks)+3] == "⚪" * len(final_output[len(blocks)+3])):
            final_output = final_output[:len(blocks)+3] + final_output[len(blocks)+4:]

        for x in final_output:
            f.write("".join(list(x)) + "\n")

File: Rust/src/piet_stack/stk_to_piet.rs
use super::Expr::{self, *};
use crate::optimize_stk::StackOptimizer;
use crate::piet_color::*;
use crate::piet_interpreter::*;
use image::Rgb;
use image::RgbImage;
use ndarray::Array;
use ndarray::Ix2;
use num::ToPrimitive;
use num::*;
use std::cmp;
use std::collections::HashMap;

impl super::PietStackExecutor {
    fn index_and_command_to_color_and_next_index(cmd: CMD, output: &mut Vec<String>) {
        let (mut previous_c, mut previous_r): (usize, usize) = REV_MAP[&output[output.len() - 1]];
        match cmd {
            CMD::Nop => {
                output.push(String::from("⚪"));
                output.push(String::from("🔴"));
                return;
            }
            CMD::Push(i) => {
                let x: BigInt = i - Into::<BigInt>::into(1);
                for _ in 0..(x.to_usize().unwrap()) {
                    output.push(String::from(COLORS[previous_r][previous_c]));
                }
                previous_r = (previous_r + 1) % 3
            }
            CMD::Pop => {
                previous_r = (previous_r + 2) % 3;
            }
            CMD::Add => {
                previous_c = (previous_c + 1) % 6;
            }
            CMD::Sub => {
                previous_c = (previous_c + 1) % 6;
                previous_r = (previous_r + 1) % 3;
            }
            CMD::Mul => {
                previous_c = (previous_c + 1) % 6;
                previous_r = (previous_r + 2) % 3;
            }
            CMD::Div => {
                previous_c = (previous_c + 2) % 6;
            }
            CMD::Mod => {
                previous_c = (previous_c + 2) % 6;
                previous_r = (previous_r + 1) % 3;
            }
            CMD::Not => {
                previous_c = (previous_c + 2) % 6;
                previous_r = (previous_r + 2) % 3;
            }
            CMD::Greater => {
                previous_c = (previous_c + 3) % 6;
            }
            CMD::Pointer => {
                previous_c = (previous_c + 3) % 6;
                previous_r = (previous_r + 1) % 3;
            }
            CMD::Switch => {
                previous_c = (previous_c + 3) % 6;
                previous_r = (previous_r + 2) % 3;
            }
            CMD::Dup => {
                previous_c = (previous_c + 4) % 6;
            }
            CMD::Roll => {
                previous_c = (previous_c + 4) % 6;
                previous_r = (previous_r + 1) % 3;
            }
            CMD::InN => {
                previous_c = (previous_c + 4) % 6;
                previous_r = (previous_r + 2) % 3;
            }
            CMD::InC => {
                previous_c = (previous_c + 5) % 6;
            }
            CMD::OutN => {
                previous_c = (previous_c + 5) % 6;
                previous_r = (previous_r + 1) % 3;
            }
            CMD::OutC => {
                previous_c = (previous_c + 5) % 6;
                previous_r = (previous_r + 2) % 3;
            }
        }
        output.push(String::from(COLORS[previous_r][previous_c]))
    }

    fn make_block(block: Vec<Expr>) -> (Vec<String>, (Option<String>, Option<String>, bool)) {
        let mut output = Vec::new();
        output.push(String::from("🔴"));
        for inp in block {
            match inp {
                Branch(t, e) => return (output, (Some(t), Some(e), false)),
                Goto(l) => return (output, (Some(l), None, false)),
                // Term =>
                //     return (output, (Some("term"), None)),
                Instr(cmd) => {
                    Self::index_and_command_to_color_and_next_index(cmd, &mut output);
                }
                Debug => {}
                Comment(_) => {}
                GotoStk => {
                    return (output, (None, None, true))
                    // output.push(String::from("⚫"));
                    // return (output, (Some(String::from("term")), None));
                    // NOP?
                    // todo!("goto_stk")
                }
            }
        }
        return (output, (None, None, false));
    }

    fn split_in_blocks(
        to_split: Vec<String>,
        j_width: usize,
        going_right: bool,
    ) -> (Vec<Vec<String>>, bool) {
        let mut block_blocks = vec![];
        let mut offset = 0;
        let mut running = true;
        let mut going_right = !going_right;
        while running {
            let last_offset = offset;
            let mut broke = false;
            for j in 0..j_width {
                if offset + j >= to_split.len() {
                    offset = to_split.len();
                    running = false;
                    broke = true;
                    break;
                }
                // TODO: What does this do?
                if offset + j_width < to_split.len() {
                    let mut typ_of_elems = vec![];
                    for i in to_split[offset + j..offset + j_width].into_iter() {
                        let mut do_break = true;
                        for ind in typ_of_elems.clone() {
                            if i == ind {
                                do_break = false;
                                break;
                            }
                        }
                        if do_break {
                            typ_of_elems.push(i);
                        }
                    }
                    if typ_of_elems.len() == 2 {
                        if j == 0 {
                            panic!("j was zero");
                        }
                        offset += j;
                        broke = true;
                        break;
                    }
                }
            }
            if !broke {
                offset += j_width
            }
            block_blocks.push(
                to_split
                    [last_offset..cmp::min(offset + (if running { 1 } else { 0 }), to_split.len())]
                    .to_vec(),
            );
            going_right = !going_right;
        }
        return (block_blocks, going_right);
    }

    fn goto_block_coord(optimizer: &mut StackOptimizer, id: usize) -> Vec<Expr> {
        let mut goto_exprs: Vec<Expr> = vec![];
        goto_exprs.extend(
            optimizer
                .optimize_number(id.into())
                .into_iter()
                .map(Instr)
                .collect::<Vec<Expr>>(),
        );
        goto_exprs
    }

    pub fn to_png(&self, optimizer: &mut StackOptimizer) -> image::RgbImage {
        let (parsed_blocks, block_index) = (self.blocks.clone(), self.block_index.clone());

        let pbl: usize = parsed_blocks.len() + 1;
        let b_width = (pbl as f32).sqrt().ceil() as usize; // (pbl).isqrt();
        let b_height = (pbl - 1) / b_width + 1;

        // println!("Blocks");

        let mut blocks: HashMap<String, (Vec<String>, (Option<String>, Option<String>, bool))> =
            HashMap::new();

        for (k, b) in parsed_blocks {
            // println!("Blocks: {}, {:?}", k, b);

            let mut nb = vec![];
            for x in b {
                // println!("      : -> {:?}", x);
                match x {
                    Instr(CMD::Push(n)) if n == 0.into() => {
                        nb.push(Instr(CMD::Push(1.into())));
                        nb.push(Instr(CMD::Not));
                    }
                    Instr(CMD::Push(n)) => {
                        if n < 0.into() {
                            nb.push(Instr(CMD::Push(1.into())));
                            nb.push(Instr(CMD::Not));
                            nb.extend(
                                optimizer
                                    .optimize_number(-n)
                                    .into_iter()
                                    .map(Instr)
                                    .collect::<Vec<Expr>>(),
                            );
                            nb.push(Instr(CMD::Sub));
                        } else {
                            nb.extend(
                                optimizer
                                    .optimize_number(n)
                                    .into_iter()
                                    .map(Instr)
                                    .collect::<Vec<Expr>>(),
                            );
                        }
                    }
                    _ => nb.push(x),
                }
            }

            blocks.insert(String::from(k), Self::make_block(nb));
        }

        // println!("Parsed blocks");

        let mut mid_blocks: HashMap<
            String,
            (Vec<String>, (Option<Vec<String>>, Option<Vec<String>>)),
        > = HashMap::new();

        // println!("Mid blocks");

        for (x, (x_vec, (x_t, x_e, x_g))) in blocks {
            if x_g {
                mid_blocks.insert(x, (x_vec.clone(), (Some(vec![]), None)));
            } else if x_t.is_some() && x_e.is_none() {
                if !block_index.contains_key(&x_t.clone().unwrap()) {
                    panic!("no block with name: {}", &x_t.unwrap());
                }

                // is goto
                let goto_statement = Self::make_block(Self::goto_block_coord(
                    optimizer,
                    block_index[&x_t.unwrap()],
                ))
                .0;

                mid_blocks.insert(x, (x_vec.clone(), (Some(goto_statement), None)));
                // TODO: make goto use color, instead of white!;
            } else if x_t.is_some() && x_e.is_some() {
                let mut goto_exprs_1: Vec<Expr> = vec![];
                goto_exprs_1.push(Instr(CMD::Pointer));
                goto_exprs_1.push(Instr(CMD::Nop));
                goto_exprs_1.extend(Self::goto_block_coord(
                    optimizer,
                    block_index[&x_t.unwrap()],
                ));

                // output.push("🔴")
                let goto_statement_1: Vec<_> = Self::make_block(goto_exprs_1).0;

                let mut goto_exprs_2: Vec<Expr> = vec![];
                goto_exprs_2.push(Instr(CMD::Pointer));
                goto_exprs_2.push(Instr(CMD::Nop));
                goto_exprs_2.extend(Self::goto_block_coord(
                    optimizer,
                    block_index[&x_e.unwrap()],
                ));

                let goto_statement_2: Vec<_> = Self::make_block(goto_exprs_2).0;

                mid_blocks.insert(
                    x,
                    (
                        x_vec.clone(),
                        (Some(goto_statement_1), Some(goto_statement_2)),
                    ),
                );
            } else {
                mid_blocks.insert(x, (x_vec, (None, None)));
            }
        }

        // println!("J width");

        // // # Split blocks
        let j_width = cmp::max(70, cmp::max(b_width, b_height) / 5);
        // let j_width = cmp::max(30, cmp::max(b_width, b_height) / 5);

        let mut final_blocks: HashMap<String, (Vec<Vec<String>>, Vec<String>)> = HashMap::new();

        // println!("Final blocks");

        for (x, (bx, (bx_t, bx_e))) in mid_blocks {
            let (mut splits, going_right) = Self::split_in_blocks(bx, j_width, true);
            let mut branch_blocks = vec![];
            if bx_t.is_some() && bx_e.is_none() {
                let (goto_splits, _) = Self::split_in_blocks(bx_t.unwrap(), j_width, true);
                splits.extend(goto_splits);
            } else if bx_t.is_some() && bx_e.is_some() {
                if going_right {
                    splits.push(vec![]);
                }
                let (goto_splits, _) = Self::split_in_blocks(bx_e.unwrap(), j_width, true);
                splits.extend(goto_splits);
                branch_blocks = bx_t.unwrap();
            }
            final_blocks.insert(x, (splits, branch_blocks));
        }

        // println!("Draw?");

        Self::to_png_draw(
            optimizer,
            b_width,
            b_height,
            final_blocks,
            block_index,
            j_width,
        )
    }

    // (Vec<&'a str>, (Option<&'a str>, Option<&'a str>))

    pub fn to_png_draw(
        optimizer: &mut StackOptimizer,
        b_width: usize,
        b_height: usize,
        final_blocks: HashMap<String, (Vec<Vec<String>>, Vec<String>)>,
        block_index: HashMap<String, usize>,
        j_width: usize,
    ) -> image::RgbImage {
        let pre = ["⚪", "⚪", "⚪"];
        let post = ["⚪"];

        let total_block_width = (j_width + 4 + 1) + 6;
        let height_per_block = final_blocks
            .clone()
            .into_iter()
            .map(|(_, (bx, bx_branch))| bx.len() * 5 / 2 + bx_branch.len())
            .max()
            .unwrap();

        let total_block_height = height_per_block + 7 + 25;

        let w = pre.len() + total_block_width * b_width + post.len();
        let h = 2 + 1 + total_block_height * b_height + 2;

        let mut arr: Array<String, Ix2> = Array::default((w, h));
        for y in 0..h {
            for x in 0..w {
                arr[(x, y)] = String::from("⚪");
            }
        }

        fn set_range(arr: &mut Array<String, Ix2>, h: usize, i: usize, j: usize, c: String) {
            for x in i..(i + j) {
                arr[(x, h)] = c.clone();
            }
        }

        set_range(
            &mut arr,
            1,
            0,
            pre.len() + total_block_width * b_width,
            String::from("⚫"),
        );

        for i in 0..b_height {
            set_range(
                &mut arr,
                1 + (1 + i) * total_block_height,
                pre.len(),
                total_block_width * b_width,
                String::from("⚫"),
            );
            set_range(
                &mut arr,
                3 + i * total_block_height,
                pre.len() - 1,
                1 + total_block_width * b_width,
                String::from("⚫"),
            );
            set_range(
                &mut arr,
                2 + (1 + i) * total_block_height,
                pre.len() - 2,
                1,
                String::from("⚫"),
            );
        }

        // Goto converter to a pair of numbers
        let converter_x_offset = 8;
        for (i, v) in Self::make_block(vec![
            Instr(CMD::Dup),
            Instr(CMD::Push(b_width.into())),
            Instr(CMD::Dup),
            Instr(CMD::Push(3.into())),
            Instr(CMD::Push(1.into())),
            Instr(CMD::Roll),
            Instr(CMD::Mod),
            Instr(CMD::Push(3.into())),
            Instr(CMD::Push(1.into())),
            Instr(CMD::Roll),
            Instr(CMD::Div),
        ])
        .0
        .into_iter()
        .enumerate()
        {
            arr[(converter_x_offset + i, 0)] = v.clone();
            arr[(converter_x_offset + i, 2)] = v;
        }
        // arr[(4, 2)] = String::from("❤");

        for (i, branch_instr) in
            Self::make_block(Self::goto_block_coord(optimizer, block_index["main"]))
                .0
                .into_iter()
                .enumerate()
        {
            arr[(i, 0)] = branch_instr;
        }

        let prepare_pointer = vec![
            Instr(CMD::Dup),
            Instr(CMD::Push(1.into())),
            Instr(CMD::Not),
            Instr(CMD::Greater),
            Instr(CMD::Not),
            Instr(CMD::Pointer),
            Instr(CMD::Push(1.into())),
            Instr(CMD::Sub),
        ];
        let prepare_pointer_index = 5 + 1; // prepare_pointer.index("pointer")+1

        let prepare_pointer_block = Self::make_block(prepare_pointer.clone()).0;
        let mut temp: Vec<_> = prepare_pointer[0..prepare_pointer_index].to_vec();
        temp.push(Instr(CMD::Pop));
        let prepare_pointer_pop = Self::make_block(temp).0;
        let prepare_pointer_pop =
            &prepare_pointer_pop[prepare_pointer_index + 1..prepare_pointer_pop.len()];

        for li in 0..b_height {
            for (j, branch_instr) in prepare_pointer_block.clone().into_iter().enumerate() {
                arr[(
                    pre.len() + total_block_width * b_width,
                    1 + total_block_height * li + total_block_height - prepare_pointer_index
                        + j
                        + 1,
                )] = branch_instr;
            }
            for (j, y) in prepare_pointer_pop.into_iter().enumerate() {
                arr[(
                    pre.len() + total_block_width * b_width - 1 - j,
                    1 + total_block_height * li + total_block_height + 1,
                )] = String::from(y.clone());
            }
        }

        fn block_line_left(arr: &mut Array<String, Ix2>, yi: usize, xi: usize, gap: usize) {
            // gap = 2
            arr[(xi - 3, yi)] = String::from("⚫");
            arr[(xi - 2, yi - 2)] = String::from("⚫");
            arr[(xi - 1, yi - 1)] = String::from("⚫");
            arr[(xi - 2, yi + gap + 2)] = String::from("⚫");
            arr[(xi - 4, yi + gap + 1)] = String::from("⚫");
            arr[(xi - 3, yi + gap)] = String::from("⚫");
        }

        fn block_line_right(arr: &mut Array<String, Ix2>, yi: usize, xi: usize, gap: usize) {
            // gap = 2
            arr[(xi + 1, yi)] = String::from("⚫");
            arr[(xi, yi + gap + 2)] = String::from("⚫");
        }

        // Draw all the code blocks
        for (x, (bx, bx_branch)) in final_blocks {
            let si = block_index[&x];

            let start_index = 7 + pre.len() + total_block_width * (si % b_width);
            let mut li = 7 + total_block_height * (si / b_width); // line index

            for (j, y) in prepare_pointer_block.clone().into_iter().enumerate() {
                arr[(start_index - prepare_pointer_index - 1 + j, li - 3)] =
                    String::from(y.clone());
            }
            for (j, y) in prepare_pointer_pop.into_iter().enumerate() {
                arr[(start_index - 1, li - 3 + 1 + j)] = String::from(y.clone());
            }

            if x == "term" {
                let (heart_x, heart_y) = (li + 1, start_index - 1);

                arr[(heart_y - 1, heart_x)] = String::from("🔴");
                arr[(heart_y, heart_x)] = String::from("🔴");
                arr[(heart_y + 1, heart_x)] = String::from("🔴");
                arr[(heart_y, heart_x + 1)] = String::from("🔴");

                arr[(heart_y - 1, heart_x - 1)] = String::from("⚫");
                arr[(heart_y + 1, heart_x - 1)] = String::from("⚫");
                arr[(heart_y - 2, heart_x)] = String::from("⚫");
                arr[(heart_y + 2, heart_x)] = String::from("⚫");
                arr[(heart_y - 1, heart_x + 1)] = String::from("⚫");
                arr[(heart_y + 1, heart_x + 1)] = String::from("⚫");
                arr[(heart_y, heart_x + 2)] = String::from("⚫");

                continue;
            }

            arr[(start_index - 1, li + 1)] = String::from("⚫");
            arr[(start_index - 3, li)] = String::from("⚫");
            arr[(start_index - 2, li - 1)] = String::from("⚫");

            let right_line_gap = 2;
            let left_line_gap = 1;

            let mut going_right = true;
            for block_block in bx {
                if going_right {
                    for (j, y) in block_block.into_iter().enumerate() {
                        arr[(start_index + j, li)] = String::from(y.clone());
                    }
                    block_line_right(&mut arr, li, start_index + j_width, right_line_gap);
                    li = li + right_line_gap + 1;
                } else {
                    // going left
                    for (j, y) in block_block.into_iter().enumerate() {
                        arr[(j_width + start_index - 1 - j, li)] = String::from(y.clone());
                    }
                    block_line_left(&mut arr, li, start_index, left_line_gap);
                    li = li + left_line_gap + 1;
                }
                going_right = !going_right;
            }

            if bx_branch.len() > 0 {
                for (j, y) in bx_branch[2..].into_iter().enumerate() {
                    arr[(start_index + 1, li - right_line_gap + j)] = String::from(y.clone());
                }
            }

            if !going_right {
                arr[(start_index + j_width, li + 1)] = String::from("⚪"); // 🔴
            } else {
                arr[(start_index - 2, li + 1)] = String::from("⚪"); // 🔴
            }
        }

        let mut new_h = 1;
        let mut remove_indexes = vec![];
        for y in 0..h - 1 {
            let mut success = true;
            for x in 0..w {
                if arr[(x, y)] != String::from("⚪") || arr[(x, y + 1)] != String::from("⚪") {
                    success = false;
                    break;
                }
            }
            if success {
                remove_indexes.push(y);
            } else {
                new_h += 1;
            }
        }

        let mut fin_arr: Array<String, Ix2> = Array::default((w, new_h));
        let mut h_ = 0;
        for y in 0..h {
            if remove_indexes.contains(&y) {
                h_ += 1;
                continue;
            }
            for x in 0..w {
                fin_arr[(x, y - h_)] = arr[(x, y)].clone();
            }
        }

        let mut img = RgbImage::new(w as u32, new_h as u32);

        for y in 0..new_h {
            for x in 0..w {
                let c: ValidColor = fin_arr[(x, y)].as_str().into();
                let r: Rgb<u8> = c.into();
                img[(x as u32, y as u32)] = r;
            }
        }

        return img;
    }
}
